File: 백준/Gold/4781. 사탕 가게/사탕 가게.py
import sys

# 빠른 입력을 위해 sys.stdin.readline 사용
input = sys.stdin.readline

results = [] # 각 테스트 케이스 결과를 저장할 리스트

while True:
    line = input().split()
    n = int(line[0])
    # 입력이 없거나 빈 줄이면 종료 (EOF 처리)
    if not line or n == 0:
        # m 값도 확인하여 0.00인지 보는 것이 더 안전할 수 있으나,
        # 문제 조건상 n=0이면 m=0.00이므로 n만 확인
        try:
            m_double = float(line[1])
            # 부동소수점 비교는 오차 감안 (epsilon 사용)
            # epsilon = 1e-9
            # if n == 0 and abs(m_double - 0.00) < epsilon:
            #     break
            # 문제 조건상 n=0이면 m=0.00이므로 아래 break로 충분
            if n == 0:
                 break
        except (IndexError, ValueError): # n만 입력되거나 잘못된 입력 처리
             break
        # 정상적인 종료 조건이 아니면 계속 진행 (혹은 에러 처리)
        # 여기서는 n=0이면 정상 종료로 간주

    m_double = float(line[1])

    # 1. 예산을 센트 단위의 정수로 변환
    m_cents = int(m_double * 100 + 0.5)

    # 2. DP 배열 선언 및 초기화
    # dp[j] = 예산 j 센트로 얻을 수 있는 최대 칼로리
    dp = [0] * (m_cents + 1) # 크기는 m_cents + 1 (0부터 m_cents까지)

    candies = [] # 사탕 정보를 저장할 리스트
    for _ in range(n):
        c_line = input().split()
        calories = int(c_line[0])
        price_double = float(c_line[1])
        # 가격을 센트 단위의 정수로 변환
        price_cents = int(price_double * 100 + 0.5)
        candies.append((calories, price_cents))

    # 3 & 4. DP 테이블 갱신 (무한 배낭 문제)
    for calories, price_cents in candies:
        # 가격이 0인 사탕은 예산 내에서 무한히 먹을 수 있으나,
        # 문제 제약상 가격은 0.01 이상일 것으로 예상됨.
        # 만약 가격이 0이고 칼로리가 양수면 답은 무한대가 될 수 있음.
        # 하지만 이 문제에서는 그런 경우는 없는 것으로 보임.
        if price_cents == 0 and calories > 0:
             # 특별 처리 필요 (문제 조건 확인 필요)
             # 여기서는 가격이 항상 0보다 크다고 가정
             continue

        # j는 현재 사탕의 가격(센트)부터 시작
        # price_cents가 0인 경우 range 시작점에 주의해야 하나, 위에서 처리함
        for j in range(price_cents, m_cents + 1):
            # 점화식 적용
            dp[j] = max(dp[j], dp[j - price_cents] + calories)

    # 5. 최종 결과 저장
    results.append(dp[m_cents])

# 결과 한 번에 출력
for result in results:
    print(result)

File: 백준/Gold/4781. 사탕 가게/사탕 가게.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        while (true) {
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken()); // 사탕 종류의 수
            double mDouble = Double.parseDouble(st.nextToken()); // 예산 (달러)

            // 종료 조건 확인 (n=0, m=0.00)
            // double 비교는 오차가 있을 수 있으므로 n=0으로 주로 확인하고,
            // mDouble이 0에 매우 가까운지 확인하는 보조적인 방법을 사용할 수 있습니다.
            // 여기서는 n=0이면 m도 0.00이라고 문제에서 명시했으므로 n만 확인해도 충분합니다.
            if (n == 0) {
                // mDouble이 0.00인지 확실히 하기 위한 추가 검사 (선택적)
                // double 비교는 epsilon을 사용하는 것이 안전합니다.
                // double epsilon = 1e-9;
                // if (Math.abs(mDouble - 0.00) < epsilon) {
                //     break;
                // }
                // 문제 조건상 n=0이면 m=0.00이므로 아래 break만으로 충분
                 break;
            }

            // 1. 예산을 센트 단위의 정수로 변환
            int mCents = (int) (mDouble * 100 + 0.5); // 부동소수점 오차 보정 및 반올림

            // 2. DP 배열 선언 및 초기화
            // dp[j] = 예산 j 센트로 얻을 수 있는 최대 칼로리
            int[] dp = new int[mCents + 1]; // 크기는 mCents + 1 (0부터 mCents까지)
            // Java에서 int 배열은 자동으로 0으로 초기화됩니다.

            // 3 & 4. 각 사탕 정보 입력 및 DP 테이블 갱신
            for (int i = 0; i < n; i++) {
                st = new StringTokenizer(br.readLine());
                int calories = Integer.parseInt(st.nextToken()); // 칼로리
                double priceDouble = Double.parseDouble(st.nextToken()); // 가격 (달러)

                // 가격을 센트 단위의 정수로 변환
                int priceCents = (int) (priceDouble * 100 + 0.5);

                // 4. DP 점화식 적용 (무한 배낭 문제)
                // j는 현재 사탕의 가격(센트)부터 시작해야 함
                for (int j = priceCents; j <= mCents; j++) {
                    // 현재 예산 j에서
                    // 1. 이 사탕을 사지 않는 경우: dp[j] (이전 값 유지)
                    // 2. 이 사탕을 사는 경우: dp[j - priceCents] + calories
                    // 둘 중 더 큰 값으로 갱신
                    dp[j] = Math.max(dp[j], dp[j - priceCents] + calories);
                }
            }

            // 5. 최종 결과 저장
            sb.append(dp[mCents]).append("\n");
        }

        // 결과 출력
        System.out.print(sb.toString());
        br.close();
    }
}